app/components-react/windows/settings/Hotkey.tsx

```diff
@@ -0,0 +1,216 @@
+    return newBindings;
+  };
+}
+
+function removeBindingAt(index: number) {
+  return (bindings: IKeyedBinding[]) => {
+    // If this is the last binding, replace it with an
+    // empty binding instead.
+    if (bindings.length === 1) {
+      bindings[0].binding = getBlankBinding();
+    } else {
+      bindings.splice(index, 1);
+    }
+
+    return bindings;
+  };
+}
+
+function setBindingAtIndex(index: number, binding: IBinding) {
+  return (bindings: IKeyedBinding[]) => {
+    const newBindings = [...bindings];
+    newBindings.splice(index, 1, createBindingWithKey(binding));
+    return newBindings;
+  };
+}
+
+/**
+ * Turns a binding into a string representation
+ */
+function getBindingString(binding: IBinding) {
+  const keys: string[] = [];
+
+  if (binding.modifiers.alt) keys.push(byOS({ [OS.Windows]: 'Alt', [OS.Mac]: 'Opt' }));
+  if (binding.modifiers.ctrl) keys.push('Ctrl');
+  if (binding.modifiers.shift) keys.push('Shift');
+  if (binding.modifiers.meta) keys.push(byOS({ [OS.Windows]: 'Win', [OS.Mac]: 'Cmd' }));
+
+  let key = binding.key;
+
+  const matchDigit = binding.key.match(/^Digit([0-9])$/);
+  if (matchDigit) key = matchDigit[1];
+
+  const matchKey = binding.key.match(/^Key([A-Z])$/);
+  if (matchKey) key = matchKey[1];
+
+  if (key === 'MiddleMouseButton') key = 'Mouse 3';
+  if (key === 'X1MouseButton') key = 'Mouse 4';
+  if (key === 'X2MouseButton') key = 'Mouse 5';
+
+  keys.push(key);
+
+  return keys.join('+');
+}
+
+export default function Hotkey(props: HotkeyProps) {
+  const { MarkersService } = Services;
+  const { hotkey } = props;
+  const description = hotkey.description!;
+
+  const [bindings, setBindings] = useState<IKeyedBinding[]>(
+    hotkey.bindings.map(createBindingWithKey),
+  );
+
+  /**
+   * Adds a new blank binding
+   */
+  function addBinding(index: number) {
+    return setBindings(addEmptyBindingAt(index));
+  }
+
+  function removeBinding(index: number) {
+    return setBindings(removeBindingAt(index));
+  }
+
+  /**
+   * Sets the bindings on the hotkey object
+   */
+  function setupBindings(bindingList = bindings) {
+    const newBindings = bindingList.reduce((acc, binding) => {
+      if (binding.binding.key) {
+        acc.push(binding);
+      }
+
+      return acc;
+    }, [] as IKeyedBinding[]);
+
+    // TODO: avoid mutation
+    hotkey.bindings = newBindings.map(b => b.binding);
+
+    setBindings(newBindings);
+  }
+
+  function getMarkerValue(): string {
+    if (!hotkey.isMarker) return '';
+    return MarkersService.views.getLabel(hotkey.actionName);
+  }
+
+  function setMarkerValue(value: string): void {
+    if (!hotkey.isMarker) return;
+    MarkersService.actions.setMarkerName(hotkey.actionName, value);
+  }
+
+  useEffect(() => {
+    if (bindings.length === 0) {
+      setBindings([createBindingWithKey(getBlankBinding())]);
+    }
+  }, []);
+
+  // TODO: avoid mutation
+  useEffect(() => {
+    hotkey.bindings = bindings.map(b => b.binding);
+  }, [bindings]);
```

Comment:
i'm not 100% sure this is going to work as expected, as i don't think React will call this when there is a deep change in the bindings array